File: src/credentials/pass.rs
//! Password-store (pass) credential backend.
//!
//! Retrieves credentials from pass entries. Each entry can contain multiple
//! fields in the format `field-name: value`.

use std::collections::HashMap;
use std::process::Command;

use anyhow::{Context, Result};
use async_trait::async_trait;
use secrecy::{ExposeSecret, SecretString};
use serde::{Deserialize, Serialize};

use super::field_entry::FieldEntry;
use super::CredentialStore;

/// Configuration for a pass credential store.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PassConfig {
    /// The pass entry path (e.g., "finance/coinbase-api").
    pub path: String,

    /// Mapping from logical key names to field names in the pass entry.
    /// If not specified, the logical key name is used as-is.
    #[serde(default)]
    pub fields: HashMap<String, String>,
}

/// Credential store backed by password-store (pass).
///
/// Reads credentials from a pass entry, parsing fields in the format:
/// ```text
/// field-name: value
/// ```
///
/// The first line of the entry is treated as the "password" field.
pub struct PassCredentialStore {
    config: PassConfig,
}

impl PassCredentialStore {
    /// Create a new pass credential store with the given configuration.
    pub fn new(config: PassConfig) -> Self {
        Self { config }
    }

    /// Create a store for a simple pass entry path, using key names directly as field names.
    pub fn from_path(path: impl Into<String>) -> Self {
        Self::new(PassConfig {
            path: path.into(),
            fields: HashMap::new(),
        })
    }

    /// Get the field name in the pass entry for a logical key.
    fn field_name<'a>(&'a self, key: &'a str) -> &'a str {
        self.config
            .fields
            .get(key)
            .map(|s| s.as_str())
            .unwrap_or(key)
    }

    /// Read and parse the pass entry.
    fn read_entry(&self) -> Result<FieldEntry> {
        let output = Command::new("pass")
            .arg("show")
            .arg(&self.config.path)
            .output()
            .context("Failed to run pass command")?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("pass command failed: {}", stderr.trim());
        }

        let content = String::from_utf8(output.stdout).context("Invalid UTF-8 in pass output")?;

        Ok(FieldEntry::parse(&content))
    }

    /// Write a field to the pass entry.
    fn write_field(&self, field: &str, value: &str) -> Result<()> {
        // Read existing entry
        let mut entry = self.read_entry().unwrap_or_default();

        // Update the field
        entry.fields.insert(field.to_string(), value.to_string());

        // Reconstruct the entry content
        let content = entry.to_string();

        // Write back using pass insert
        let mut child = Command::new("pass")
            .arg("insert")
            .arg("--multiline")
            .arg("--force")
            .arg(&self.config.path)
            .stdin(std::process::Stdio::piped())
            .spawn()
            .context("Failed to spawn pass command")?;

        use std::io::Write;
        if let Some(stdin) = child.stdin.as_mut() {
            stdin
                .write_all(content.as_bytes())
                .context("Failed to write to pass stdin")?;
        }

        let status = child.wait().context("Failed to wait for pass command")?;
        if !status.success() {
            anyhow::bail!("pass insert command failed");
        }

        Ok(())
    }
}

#[async_trait]
impl CredentialStore for PassCredentialStore {
    async fn get(&self, key: &str) -> Result<Option<SecretString>> {
        let field = self.field_name(key);
        let entry = self.read_entry()?;

        Ok(entry
            .fields
            .get(field)
            .map(|v| SecretString::from(v.clone())))
    }

    async fn set(&self, key: &str, value: SecretString) -> Result<()> {
        let field = self.field_name(key);
        self.write_field(field, value.expose_secret())?;
        Ok(())
    }
}

#[cfg(test)]
#[path = "../../tests/unit/credentials/pass_tests.rs"]
mod pass_tests;

File: ts/src/credentials/pass.ts
import { execFile, spawn } from 'node:child_process';
import { promisify } from 'node:util';

import type { CredentialStore } from './credential-store.js';
import type { PassConfig } from './credential-config.js';

const execFileAsync = promisify(execFile);

type PassEntry = {
  password: string | null;
  fields: Map<string, string>;
};

function parsePassEntry(content: string): PassEntry {
  const lines = content.split(/\r?\n/);
  const password = lines.length > 0 && lines[0] !== '' ? lines[0] : null;
  const fields = new Map<string, string>();

  if (password !== null) {
    fields.set('password', password);
  }

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    const idx = line.indexOf(': ');
    if (idx === -1) continue;
    const key = line.slice(0, idx);
    const value = line.slice(idx + 2).replace(/\\n/g, '\n');
    fields.set(key, value);
  }

  return { password, fields };
}

function serializePassEntry(entry: PassEntry): string {
  const lines: string[] = [];
  lines.push(entry.password ?? '');

  const keys = Array.from(entry.fields.keys()).filter((k) => k !== 'password');
  keys.sort((a, b) => a.localeCompare(b));
  for (const key of keys) {
    const value = entry.fields.get(key);
    if (value === undefined) continue;
    lines.push(`${key}: ${value.replace(/\n/g, '\\n')}`);
  }

  return lines.join('\n') + '\n';
}

/**
 * Credential store backed by password-store (pass).
 *
 * Mirrors the Rust `PassCredentialStore` behavior:
 * - `pass show <path>` reads the entry.
 * - First line is treated as "password" and also available under key "password".
 * - Other lines parsed as `field: value`, with `\\n` unescaped in values.
 */
export class PassCredentialStore implements CredentialStore {
  private readonly config: PassConfig;

  constructor(config: PassConfig) {
    this.config = config;
  }

  private fieldName(key: string): string {
    return this.config.fields[key] ?? key;
  }

  private async readEntry(): Promise<PassEntry> {
    const { stdout } = await execFileAsync('pass', ['show', this.config.path], {
      encoding: 'utf8',
      maxBuffer: 10 * 1024 * 1024,
    });
    return parsePassEntry(stdout);
  }

  private async writeEntry(entry: PassEntry): Promise<void> {
    const content = serializePassEntry(entry);
    await new Promise<void>((resolve, reject) => {
      const child = spawn('pass', ['insert', '--multiline', '--force', this.config.path], {
        stdio: ['pipe', 'ignore', 'pipe'],
      });

      let stderr = '';
      child.stderr.setEncoding('utf8');
      child.stderr.on('data', (chunk) => {
        stderr += chunk;
      });

      child.on('error', (err) => reject(err));
      child.on('close', (code) => {
        if (code === 0) resolve();
        reject(new Error(`pass insert failed (code ${code}): ${stderr.trim()}`));
      });

      child.stdin.setDefaultEncoding('utf8');
      child.stdin.write(content);
      child.stdin.end();
    });
  }

  async get(key: string): Promise<string | null> {
    const field = this.fieldName(key);
    const entry = await this.readEntry();
    return entry.fields.get(field) ?? null;
  }

  async set(key: string, value: string): Promise<void> {
    const field = this.fieldName(key);

    let entry: PassEntry;
    try {
      entry = await this.readEntry();
    } catch {
      entry = { password: null, fields: new Map<string, string>() };
    }

    // Maintain password line separately if caller sets "password".
    if (field === 'password') {
      entry.password = value;
      entry.fields.set('password', value);
    } else {
      entry.fields.set(field, value);
    }

    await this.writeEntry(entry);
  }

  supportsWrite(): boolean {
    return true;
  }
}

